doc(tutorial): made a tutorial for unity developer a better understan… (#79)
Terasology/CoreRendering: docs/tutorial-for-unity.md

# Terminology for Unity Developers

## What is an FBO?

FBO stands for frame object buffer. It's equal to the renderer target that we generated.

## What is a node?

You can regard a node as a "Render Pass" in Unity.

> A node is customized code you can do whaterver you want to the render result, you can change the output of **InitialPostProcessingNode** like Tinting,

Reviewer: ```suggestion
A node is custom code that will be applied to the render result.
For example, you can change the output of **InitialPostProcessingNode** to achieve a tinting effect.
```

you can also change the input of "HighPassNode" to make everying "blooming".

> you can also create a node rendering character's outline just outline, add to then combine it with  **InitialPostProcessingNode**'s output to create an XRay node.

Reviewer: ```suggestion
You can also change the input of "HighPassNode" to apply a "blooming" effect or create a node rendering only a character's outline and combine it with the output of **InitialPostProcessingNode** to create an XRay node.
```

> > this requires technology to rendereing specificed layer, I'm not sure whether Terasology has this feature, hopefully it would add the "layer" conecpt in the future.

Reviewer: ```suggestion
_Note: This requires technology for a rendering-specific layer. Currently, it's unclear, whether Terasology supports this kind of  "layer" concept._
```

## How nodes work

a node has Inputs and Outputs, it's input are result from previous node, and it's output can be next node's input.

> For example, highPass node use `_SceneColor` as input, then bloom node use both `_SceneColor` and highpass node's output as input.

Reviewer: ```suggestion
For example, a highPass node uses `_SceneColor` as input, then a bloom node uses both `_SceneColor` and the highpass node's output as input.
```

> `_SceneColor` is the scene texture, I'm writing this style for better understanding for unity developers.

Reviewer: ```suggestion
The Unity equivalent of `_SceneColor` is the scene texture.
```

> highpass node is a node to grab light area then send to blur node to make the bright area of bloom.

Reviewer: ```suggestion
The highpass node is used to grab a light area and send it to the blur node to achieve the bright area of a bloom effect.
```

> it's being used for **InitialPostProcessingNode** as input for final FBO which is the final output of the camera.

Reviewer: ```suggestion
It's being used for **InitialPostProcessingNode** as the input for the final FBO, which is the final output of the camera.
```

## How to create a FBO

in Unity we can use the code below 

```C#
RenderTexture  computeTex = new RenderTexture (IMG_WIDTH, IMG_HEIGHT, 24);

```

in Terasology we need to setup these at `setDependencies(Context context)`

```java
//the root Fbo for everything
DisplayResolutionDependentFbo displayResolutionDependentFBOs = context.get(DisplayResolutionDependentFbo.class);
//create the fbo with scale
FBO someFbo = requiresFbo(new FboConfig(FBO_URI, HALF_SCALE, FBO.Type.DEFAULT), displayResolutionDependentFBOs);
//create the fbo with width and height
FBO someFbo2 = requiresFbo(new FboConfig(FBO_URI, IMG_WIDTH, IMG_HEIGHT, FBO.Type.NO_COLOR), displayResolutionDependentFBOs);

> ```

Reviewer: Setting up FBOs in Terasology translates to Unity as setting up rendering targets:

```C#
RenderTexture  computeTex = new RenderTexture (IMG_WIDTH, IMG_HEIGHT, 24);
```

In Terasology, FBOs can be set up using `setDependencies(Context context)`:

```java
//the root Fbo for everything
DisplayResolutionDependentFbo displayResolutionDependentFBOs = context.get(DisplayResolutionDependentFbo.class);
//create the fbo with scale
FBO someFbo = requiresFbo(new FboConfig(FBO_URI, HALF_SCALE, FBO.Type.DEFAULT), displayResolutionDependentFBOs);
//create the fbo with width and height
FBO someFbo2 = requiresFbo(new FboConfig(FBO_URI, IMG_WIDTH, IMG_HEIGHT, FBO.Type.NO_COLOR), displayResolutionDependentFBOs);
```

## Where are FBO being processed?

The FBO is being process inside the process method of a node
1. Fetch the material by `var mat = getMaterial(MATERIAL_URN)`
2. Set up material properties at `process()`
3. Render the quad

### What is the quad? Where Did i used the material?
Well, a quad is a quad that you created at ctor
and you need to activate the material in `setDependencies` like this:

```java
addDesiredStateChange(new EnableMaterial(MATERIAL_URN));
```

If you study the code deeply, you'll find that, `setDependencies(Content ctx)` is setting up OpenGL contents and then `process()` is the actual rendering logic: they are just OpenGL code wrapped into java code

## Common Issues

> If you wrote a shader, but Terasology doesn't recognize it, you must remember to create a material.

Reviewer: Cervator already asked about this: You mean that if Terasology doesn't recognize a shader I wrote, I probably forgot to create a material and have to create one, right?

Author: yes

> # Tutorial 4 Unity Developer

Reviewer 2: ```suggestion
# Terminology for Unity Developers
```
I see this as less of a tutorial (which explains how to do things) and more of an explanation about how to translate Terasology graphics terminology if you are already familiar with Unity.

## What is an FBO?

> FBO is frame object buffer, it's equal to the renderer target that we generated.

Reviewer 2: ```suggestion
FBO stands for frame object buffer. It's equal to the renderer target that we generated.
```

`FBO` is an acronym. When expanding an acronym, you would usually say that it "stands for" something or "means" something.

## What is a node?

You can regard a node as a "Render Pass" or simply a "Render Feature" in Unity.
A pass is a rendering stage, a render feature is a customized code you can do in during a stage.

> So a Terasology node is more like a "Render Feature" in Unity.

Reviewer 2: It is a "Render Pass" or a "Render Feature" then? This paragraph is a bit confusing. Are you trying to suggest that a node actually performs aspects of both things?

Author: oh, my misteake, a render feature already included render pass, render feature is like a wrapper for pass. sorry.

> Image, there is not rendering stage concept, event the rendering stage itself is a rendering feature.

Reviewer 2: There appear to be a few typos here. Did you mean this?
```suggestion
Imagine that there is no rendering stage concept. That even the rendering stage itself is a rendering feature.
```

and the rendering queue is depended on the node's connection.

> ## How are node work

Reviewer 2: ```suggestion
## How nodes work
```
It is best not to combine plural and singular ("are" is used with plurals but "node" is singular).

> a node has Inputs and Outputs, it's input are result from previous node, and it's output can be next node's input.

Reviewer 2: ```suggestion
A node has Inputs and Outputs. Its inputs are results from previous nodes, and its outputs can be next node's input.
```

"It's" is a contraction which expands to "it is". In this case I think you mean to use the possessive "its" (belongs to it). It's a very common mistake.

For example, highPass node use `_SceneColor` as input, then bloom node use both `_SceneColor` and highpass node's output as input.

`_SceneColor` is the scene texture, I'm writing this style for better understanding for unity developers.
highpass node is a node to grab light area then send to blur node to make the bright area of bloom.

> it's being used for InitialPostProcessingNode as input for final FBO which is output of the camera.

Reviewer 2: ```suggestion
It's being used by InitialPostProcessingNode as input to the final FBO which is the output of the camera.
```
I think this is what you mean but I am not sure if this correction distorts the original meaning.

## How to create a FBO

in Unity we can use the code below 

```C#
RenderTexture  computeTex = new RenderTexture (IMG_WIDTH, IMG_HEIGHT, 24);

```

in Terasology we need to setup these at `setDependencies(Context context)`

```java
//the root Fbo for everything
DisplayResolutionDependentFbo displayResolutionDependentFBOs = context.get(DisplayResolutionDependentFbo.class);
//create the fbo with scale
FBO someFbo = requiresFbo(new FboConfig(FBO_URI, HALF_SCALE, FBO.Type.DEFAULT), displayResolutionDependentFBOs);
//create the fbo with width and height
FBO someFbo2 = requiresFbo(new FboConfig(FBO_URI, IMG_WIDTH, IMG_HEIGHT, FBO.Type.NO_COLOR), displayResolutionDependentFBOs);
```

> ## Where are FBO being processed?

Reviewer 2: ```suggestion
## Where is the FBO being processed?
```

`FBO` means `Frame Buffer Object` (singular).

> The FBO is being process inside the process method of a node

Reviewer 2: ```suggestion
The FBO is being processed inside the node's process method:
```

> 1. Fetch the material by `var mat = getMaterial(MATERIAL_URN)`

Reviewer 2: The use of `var` maybe misleading here. I would rather the line specified the actual type (`Material`?). I understand this may be shorthand but examples in the documentation should be understandable in the context of the code.

2. Set up material properties at `process()`
3. Render the quad

> ### What is the quad? Where Did i used the material?

Reviewer 2: ```suggestion
### What is the quad? Where do I use the material?
```

> Well, a quad is a quad that you created at ctor

Reviewer 2: ```suggestion
Well, a quad is a quad that you created in the constructor
```

> and you need to active the material at setDependencies like this

Reviewer 2: ```suggestion
and you need to activate the material in `setDependencies` like this:
```

```java
addDesiredStateChange(new EnableMaterial(MATERIAL_URN));
```

If you have study the code deeply, you'll find that, `setDependencies(Content ctx)` are setting up OpenGL

> Contents then `process()` is the actual rendering logic, they are just openGL code wrapped into java code

Reviewer 2: ```suggestion
If you study the code deeply, you'll find that, `setDependencies(Content ctx)` is setting up OpenGL
Contents and then `process()` is the actual rendering logic: they are just OpenGL code wrapped into java code
```

> ## Whats more

Reviewer 2: ```suggestion
## Common Issues
```
I think this subheading describes the content better.

> If you wrote a shader, but Terasology doesn't recognize it, you must forget to create a material.

Reviewer 2: ```suggestion
If you wrote a shader, but Terasology doesn't recognize it, you must remember to create a material.
```

I am rarely told to forget something. I think what you intended to say was:
```suggestion
If you wrote a shader, but Terasology doesn't recognize it, you mustn't forget to create a material.
```

Either variant I have suggested here is fine.

```java
//the root Fbo for everything
DisplayResolutionDependentFbo displayResolutionDependentFBOs = context.get(DisplayResolutionDependentFbo.class);
//create the fbo with scale
FBO someFbo = requiresFbo(new FboConfig(FBO_URI, HALF_SCALE, FBO.Type.DEFAULT), displayResolutionDependentFBOs);
//create the fbo with width and height
FBO someFbo2 = requiresFbo(new FboConfig(FBO_URI, IMG_WIDTH, IMG_HEIGHT, FBO.Type.NO_COLOR), displayResolutionDependentFBOs);
```

## Where are FBO being processed?

The FBO is being process inside the process method of a node.
1. Fetch the material by `var mat = getMaterial(MATERIAL_URN)`
2. Set up material properties at `process()`
3. Render the quad

### What is the quad? Where Did i used the material?

Well, a quad is a quad that you created at ctor
and you need to activate the material in `setDependencies` like this:

```java
addDesiredStateChange(new EnableMaterial(MATERIAL_URN));
```

If you study the code deeply, you'll find that, `setDependencies(Content ctx)` is setting up OpenGL contents and then `process()` is the actual rendering logic: they are just OpenGL code wrapped into java code.

## Common Issues

### Unrecongnized Shader

> If you wrote a shader, but Terasology doesn't recognize it, you are probablly forget to create a material.

Reviewer: ```suggestion
If you wrote a shader, but Terasology doesn't recognize it, you probably forgot to create a material.
```

1. create a material called "id.mat".  

```
{

>   "shader": "MODUEL:ID",

Reviewer: ```suggestion
  "shader": "MODULE:ID",
```

  "params": {}
}
```

2. write 2 shaders `ID_vert.glsl` and `ID_frag.glsl`.

> 3. now Terasology will recognized.

Reviewer: ```suggestion
3. now Terasology should recognize your shader.
```
